### DISCUSSION TITLE
Implications of indexing a layer store by DiffID instead of ChainID?

### DISCUSSION THREAD

**Discussion by ayazhafiz** [moby/moby#47166] [Q&A]
Hi folks,

I am trying to understand the implications of having a container management tool index an image layer store by layer DiffIDs instead of ChainIDs.

As far as I understand, all major container management tools (I've looked at Docker, Podman, containerd, and nerdctl - the last of which looks to use containerd's implementation) index image layers in the layer store by ChainID. I understand that the ChainID is useful for uniquely identifying the base filesystem of an image; however, is there an inherent reason it is not sufficient (for security reasons or otherwise) to index a layer store itself by layers' DiffID?

I ask specifically because I have a use case that requires storing many images with a number of shared layers (as determined by their diff ID), but with those layers in different orders (and hence different chain IDs). I am considering a layer store implementation that would index by diff ID to enable structural sharing, but it's unclear to me if doing so has significant consequences, or if there is a better way.

Apologies if this is the wrong forum to ask this question, or if it has been answered elsewhere. I have been unable to find any online discussion of this question specifically, but if there is a well-known answer or a better forum, I would appreciate your direction.

**Comment by neersighted** [comment]
cc @stevvooe; I can try to explain it, but I think a more first-party source might have a better layman's explanation.

**Comment by stevvooe** [comment]
We included this explanation in the [spec](https://github.com/opencontainers/image-spec/blob/8797c3fedb77200c102c816ddcd295bb1e19d3e3/config.md?plain=1#L33):

> For convenience, it is sometimes useful to refer to a stack of layers with a single identifier.
> While a layer's `DiffID` identifies a single changeset, the `ChainID` identifies the subsequent application of those changesets.
> This ensures that we have handles referring to both the layer itself, as well as the result of the application of a series of changesets.

There's more in the spec to read but I'll expand a bit here.

Well, what does that mean? Let's say you have layer diffs `A` and `B` referred to by an image `a` and an image `b`. Image `a` has layers `[A,B]` and image `b` has layers `[B,A]`. If we only refer to an unpacked layer by the top diff id, we don't include the base on which we unpacked the image, which could change the result. In our example, we might have two layer `A` with the same diff ID, since they are the same layer but the actual filesystems are completely different. In a naive implementation, if you pull `a` then `b`, you'd get `[B, A, B]`. If you pull `b` then `a`, you'd get `[A,B,A]`. Oops!

To prevent this, we index the actual layer checkouts with a content addressable identifier using the order in which they are applied. Using our example above, if you pull `a` then `b`, the `ChainID([B,A])` does not match that of `ChainID([A,B])`, so a new layer base gets created, staring with `A`, then applying `B`.

Ultimately, this allows us to both refer directly to a common snapshot between container images, while still allowing changes in higher level data that may affect the digest of an image manifest.
